src/app/controllers/errors.server.controller.ts

```diff
@@ -1,39 +1,41 @@
-'use strict'
+import _ from 'lodash'
 
-const _ = require('lodash')
+import { IMongoError } from 'src/types/error'
 
 /**
  * Default error message if no more specific error
  * @type {String}
  */
-exports.defaultErrorMessage = 'An unexpected error happened. Please try again.'
+export const defaultErrorMessage =
+  'An unexpected error happened. Please try again.'
 
 /**
  * Private helper for getMongoErrorMessage to return Mongo error in a String
- * @param  {Object} err - MongoDB error object
+ * @param  {IMongoError} err - MongoDB error object
  * @return {String} errorString - Formatted error string
  */
-const mongoDuplicateKeyError = function (err) {
+
+const mongoDuplicateKeyError = function (err: IMongoError): string {
   let errorString = ''
-  try {
-    let fieldName = err.err.substring(
+  if (err.err) {
+    const fieldName = err.err.substring(
       err.err.lastIndexOf('.$') + 2,
       err.err.lastIndexOf('_1'),
     )
     errorString =
       fieldName.charAt(0).toUpperCase() + fieldName.slice(1) + ' already exists'
-  } catch (ex) {
+  } else {
     errorString = 'Unique field already exists'
   }
   return errorString
 }
 
 /**
  * Gets Mongo error object and returns formatted String for frontend
- * @param  {Objrct} err MongoDB error object
+ * @param  {IMongoError} err? MongoDB error object
  * @return {String} message - Error message returned to frontend
  */
-exports.getMongoErrorMessage = function (err) {
+export function getMongoErrorMessage(err?: IMongoError | string): string {
```

Comment:
This whole function seems like it needs a good refactor, maybe we should try to clean up the function?

I don't think we should create the `IMongoError` interface when we already have the typings of the possible errors:

```ts
import { MongoError } from 'mongodb'
import { Error as MongooseError } from 'mongoose'

...
export const getMongoErrorMessage = (
  err?: MongoError | MongooseError | string,
): string => { ... }
```

Reply:
Suggested refactor:

```ts
export const getMongoErrorMessage = (
  err?: MongoError | MongooseError | string,
): string => {
  if (!err) {
    return ''
  }

  // Handle base Mongo engine errors.
  if (err instanceof MongoError) {
    switch (err.code) {
      case 10334: // BSONObj size invalid error
        return 'Your form is too large to be supported by the system.'
      default:
        return defaultErrorMessage
    }
  }

  // Handle mongoose errors.
  if (err instanceof MongooseError.ValidationError) {
    // Join all error messages into a single message if available.
    const joinedMessage = Object.values(err.errors)
      .map((err) => err.message)
      .join(', ')

    return joinedMessage ?? err.message ?? defaultErrorMessage
  }

  if (err instanceof MongooseError) {
    return err.message ?? defaultErrorMessage
  }

  return err ?? defaultErrorMessage
}
```